### ISSUE TITLE
SUPER ISSUE: Finalize Repository Structure & Implement Core Workflow Elements

### ISSUE THREAD

**Issue by JFRudzinski** [FAIRmat-NFDI/nomad-bot-rag-docs-discord#6]
chatGPT generated plan for reference if helpful:

# 🛠 Project Infrastructure Plan

This document outlines the proposed **scaffolding, modules, and utilities** for the `nomad-bot-rag-docs-discord` project.  
It defines the minimal backbone needed to implement our RAG pipeline, keep the repo clean, and make collaboration easy during the hackathon.

---

## 📁 Directory Layout

```

nomad-bot-rag-docs-discord/
├─ app.py                         # main launcher (imports gradio\_app)
├─ pyproject.toml                 # project definition + deps
├─ .env.example                   # model/store config (editable per user)
├─ README.md                      # project overview + quickstart
├─ Makefile                       # simple commands (setup, run, lint, eval)
├─ configs/
│  └─ default.yaml                # config file (models, chunk sizes, store)
├─ data/
│  ├─ fixtures/                   # small test samples (redacted)
│  └─ README.md                   # schema + provenance notes
├─ artifacts/
│  ├─ indexes/                    # saved FAISS/Qdrant indexes
│  ├─ logs/                       # per-run JSONL logs
│  └─ experiments.csv             # (optional) eval results tracking
├─ eval/
│  ├─ gold\_nomad.jsonl            # curated gold Q\&A
│  └─ run\_eval.py                 # evaluation script
├─ scripts/
│  ├─ bootstrap\_qdrant.sh         # spin up Qdrant docker
│  ├─ embed\_docs.sh               # CLI wrapper for embedding docs
│  └─ smoke.sh                    # smoke test: end-to-end mini run
└─ src/
└─ nomad\_ragbot/
├─ **init**.py
├─ gradio\_app.py            # demo UI (Gradio)
├─ pipeline.py              # end-to-end answer\_query()
├─ llm\_client.py            # wraps LLM calls
├─ prompts.py               # builds RAG prompt
├─ logging\_utils.py         # helper for timings/scores
├─ embeddings.py            # load/embed model + caching
├─ vector\_store.py          # FAISS or Qdrant wrapper
├─ chunking.py              # text splitters (token + heading-aware)
├─ ingest\_docs.py           # parse NOMAD docs → chunks
├─ ingest\_discord.py        # parse Discord threads → chunks
├─ embed\_index.py           # index builder (CLI)
├─ retrieve.py              # retriever module
└─ rerank.py                # reranker (cross-encoder)

````

---

### `.env.example`
```env
MODEL_EMBED=BAAI/bge-base-en-v1.5
MODEL_RERANK=BAAI/bge-reranker-base
MODEL_LLM=Qwen/Qwen2.5-7B-Instruct
VECTOR_STORE=faiss   # or qdrant
QDRANT_HOST=localhost
````

### `configs/default.yaml`

```yaml
models:
  embed: BAAI/bge-base-en-v1.5
  rerank: BAAI/bge-reranker-base
  llm: Qwen/Qwen2.5-7B-Instruct
retrieval:
  k_docs: 25
  k_discord: 25
  top_n: 8
chunking:
  target_tokens: 600
  overlap_tokens: 100
store:
  type: faiss   # or qdrant
  dim: 768
```

---

### `app.py`

```python
from src.nomad_ragbot.gradio_app import demo

if __name__ == "__main__":
    demo.launch()
```

### `src/nomad_ragbot/pipeline.py`

```python
from .retrieve import retrieve_candidates
from .rerank import rerank
from .llm_client import generate
from .prompts import build_prompt

def answer_query(query: str, k_docs=25, k_discord=25, top_n=8):
    pre = retrieve_candidates(query, k_docs=k_docs, k_discord=k_discord)
    post = rerank(query, pre, top_n=top_n)
    prompt = build_prompt(query, post)
    answer = generate(prompt)
    sources = [
        {"title": c["meta"].get("title"), "url": c["meta"].get("source_url")}
        for c in post
    ]
    return {"answer": answer, "sources": sources, "retrieved": pre, "selected": post}
```

### `src/nomad_ragbot/gradio_app.py`

```python
import gradio as gr
from .pipeline import answer_query

def ui_answer(q):
    res = answer_query(q)
    cites = "\n".join(
        f"- [{s['title']}]({s['url']})" for s in res["sources"] if s["url"]
    )
    return f"{res['answer']}\n\n**Sources**\n{cites or '- (none)'}"

demo = gr.Interface(
    fn=ui_answer,
    inputs="text",
    outputs="markdown",
    title="NOMAD RAGbot",
    examples=[
        "How do I link my NOMAD account with Helmholtz AAI?",
        "How to upload data to NOMAD?"
    ]
)
```

### `src/nomad_ragbot/embeddings.py`

```python
import os
from sentence_transformers import SentenceTransformer

_EMBED = None

def get_embedder():
    global _EMBED
    if _EMBED is None:
        model = os.getenv("MODEL_EMBED", "BAAI/bge-base-en-v1.5")
        _EMBED = SentenceTransformer(model)
    return _EMBED

def embed_texts(texts):
    emb = get_embedder()
    return emb.encode(texts, normalize_embeddings=True, convert_to_numpy=True)
```

### `src/nomad_ragbot/vector_store.py`

```python
import os
STORE = os.getenv("VECTOR_STORE", "faiss").lower()

if STORE == "qdrant":
    from qdrant_client import QdrantClient
    from qdrant_client.models import VectorParams, Distance
    _client = QdrantClient(host=os.getenv("QDRANT_HOST","localhost"), port=6333)
    DIM = int(os.getenv("EMBED_DIM","768"))

    def ensure_collection(name:str):
        _client.recreate_collection(
            name,
            vectors_config=VectorParams(size=DIM, distance=Distance.COSINE)
        )

    def upsert(name, vectors, payloads):
        _client.upsert(
            name,
            points=[
                {"id": p["id"], "vector": v.tolist(), "payload": p}
                for v, p in zip(vectors, payloads)
            ]
        )

    def search(name, qvec, top_k=25):
        res = _client.search(name, query_vector=qvec.tolist(), limit=top_k)
        return [
            {"text": r.payload["text"], "meta": r.payload, "score": r.score}
            for r in res
        ]

else:
    import faiss, numpy as np
    _indexes = {}; _payloads = {}

    def ensure_collection(name:str, dim:int=768):
        _indexes[name] = faiss.IndexFlatIP(dim)
        _payloads[name] = []

    def upsert(name, vectors, payloads):
        _indexes[name].add(np.asarray(vectors, dtype="float32"))
        _payloads[name].extend(payloads)

    def search(name, qvec, top_k=25):
        import numpy as np
        D,I = _indexes[name].search(np.asarray([qvec], dtype="float32"), top_k)
        out=[]
        for s, idx in zip(D[0], I[0]):
            p = _payloads[name][idx]
            out.append({"text": p["text"], "meta": p, "score": float(s)})
        return out
```

### `src/nomad_ragbot/retrieve.py`

```python
from .embeddings import embed_texts
from .vector_store import search

def retrieve_candidates(query, k_docs=25, k_discord=25):
    qvec = embed_texts([query])[0]
    hits_docs = search("docs", qvec, top_k=k_docs)
    hits_disc = search("discord", qvec, top_k=k_discord)
    return hits_docs + hits_disc
```

### `src/nomad_ragbot/rerank.py`

```python
from sentence_transformers import CrossEncoder

_rer = None

def _get_reranker(name="BAAI/bge-reranker-base"):
    global _rer
    if _rer is None:
        _rer = CrossEncoder(name)
    return _rer

def rerank(query, candidates, top_n=8):
    if not candidates: return []
    pairs = [(query, c["text"]) for c in candidates]
    scores = _get_reranker().predict(pairs)
    ranked = sorted(zip(candidates, scores), key=lambda x: x[1], reverse=True)
    return [c for c,s in ranked[:top_n]]
```

### `src/nomad_ragbot/prompts.py`

```python
def build_prompt(question, passages):
    context = "\n\n".join([
        f"[{i+1}] {p['text']}\nSOURCE: {p['meta'].get('source_url','')}"
        for i,p in enumerate(passages)
    ])
    return (
        "You are a NOMAD support bot. Use only the context.\n"
        "If unsure, say you don't know. Cite sources as [1], [2], etc.\n\n"
        f"Question: {question}\n\nContext:\n{context}\n\nAnswer:"
    )
```

### `src/nomad_ragbot/llm_client.py`

```python
import os
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch

_model, _tok = None, None

def _load_local():
    global _model, _tok
    name = os.getenv("MODEL_LLM", "Qwen/Qwen2.5-7B-Instruct")
    _tok = AutoTokenizer.from_pretrained(name)
    _model = AutoModelForCausalLM.from_pretrained(
        name,
        torch_dtype=torch.float16,
        device_map="auto"
    )

def generate(prompt, max_new_tokens=400):
    if _model is None: _load_local()
    inputs = _tok(prompt, return_tensors="pt").to(_model.device)
    out = _model.generate(**inputs, max_new_tokens=max_new_tokens)
    return _tok.decode(out[0], skip_special_tokens=True)
```

---

## 🧪 Evaluation Setup

* **Gold Q\&A set**: `eval/gold_nomad.jsonl` (50–100 curated queries with answers + source URLs).
* **`eval/run_eval.py`**: compute

  * Recall\@k (retriever coverage)
  * MRR (rerank quality)
  * Citation accuracy (links correct)
  * Faithfulness (groundedness check)

---

## 💻 Demo UI

* **Gradio app** (`app.py` → `src/nomad_ragbot/gradio_app.py`)

  * Input: question
  * Output: generated answer + citations (clickable links)
  * Include sample queries (docs + discord)

Run with:

```bash
python app.py
```

---

### `Makefile`

```make
setup: ## install deps
	uv sync || pip install -e .

run: ## launch demo
	python app.py

embed-docs: ## embed docs chunks
	bash scripts/embed_docs.sh

lint:
	ruff check .

smoke:
	bash scripts/smoke.sh
```

* **CI (optional)**: lint + smoke test on fixtures
* **Logging**: JSONL per query (`artifacts/logs/`), pre/post rerank scores, latencies.

---

## ✅ Deliverables (by hackathon end)

* Docs + (optional) Discord chunks JSONL
* Embedding + index build script
* Retriever + reranker pipeline
* Evaluation with metrics + failure book
* Gradio demo app with citations
